src/java.base/share/classes/jdk/internal/foreign/abi/s390/zos/ZosS390CallArranger.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2022, 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023 IBM Corp. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * ===========================================================================
+ * (c) Copyright IBM Corp. 2024, 2024 All Rights Reserved
+ * ===========================================================================
+ */
+
+package jdk.internal.foreign.abi.s390.zos;
+
+import java.lang.foreign.FunctionDescriptor;
+import jdk.internal.foreign.abi.AbstractLinker.UpcallStubFactory;
+import jdk.internal.foreign.abi.DowncallLinker;
+import jdk.internal.foreign.abi.LinkerOptions;
+import jdk.internal.foreign.abi.UpcallLinker;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodType;
+
+/**
+ * For the s390 C ABI specifically, this class uses the DowncallLinker API
+ * which is turned into a MethodHandle to invoke the native code.
+ */
+public class ZosS390CallArranger {
+
+    /* Replace DowncallLinker in OpenJDK with the implementation of DowncallLinker specific to OpenJ9. */
+    public static MethodHandle arrangeDowncall(MethodType mt, FunctionDescriptor cDesc, LinkerOptions options) {
+        return DowncallLinker.getBoundMethodHandle(mt, cDesc, options);
+    }
+
+    /* Replace UpcallLinker in OpenJDK with the implementation of UpcallLinker specific to OpenJ9. */
+    public static UpcallStubFactory arrangeUpcall(MethodType mt, FunctionDescriptor cDesc, LinkerOptions options) {
+        // return UpcallLinker.makeFactory(mt, cDesc, options);
+        throw new InternalError("arrangeUpcall is not yet implemented"); //$NON-NLS-1$
```

Comment:
Please defer the error to OpenJ9 code so we don't have to change it multiple places when upcalls are supported on z/OS.
```java
        return UpcallLinker.makeFactory(mt, cDesc, options);
```

Reply:
Moved the error to OpenJ9 at [jcl/src/java.base/share/classes/jdk/internal/foreign/abi/UpcallLinker.java](https://github.com/eclipse-openj9/openj9/pull/19384/files#diff-62c41d26e433015b55af76ffba5e995f3f78efaec76c2dac9fdb104e77d54f40).